src/toxcore/dht_new/dht_node.rs

```diff
@@ -0,0 +1,528 @@
+use toxcore::dht_new::kbucket::*;
+use toxcore::dht_new::client::*;
+use toxcore::dht_new::codec::*;
+
+/// Type representing Dht UDP packets.
+//pub type DhtUdpPacket = (SocketAddr, DhtPacket);
+/// Shorthand for the transmit half of the message channel.
+type Tx = mpsc::UnboundedSender<DhtUdpPacket>;
+
+/**
+Own DHT node data.
+
+Contains:
+
+- DHT public key
+- DHT secret key
+- Close List ([`Kbucket`] with nodes close to own DHT public key)
+- ping timeout lists ([`TimeoutQueue`])
+
+# Adding node to Close List
+
+Before a [`PackedNode`] is added to the Close List, it needs to be
+checked whether:
+
+- it can be added to [`Kbucket`] \(using [`Kbucket::can_add()`])
+- [`PackedNode`] is actually online
+
+Once the first check passes node is added to the temporary list, and
+a [`NodesRequest`] request is sent to it in order to check whether it's
+online. If the node responds correctly within [`PING_TIMEOUT`], it's
+removed from temporary list and added to the Close List.
+
+[`NodesRequest`]: ../dht/struct.NodesRequest.html
+[`Kbucket`]: ../dht/struct.Kbucket.html
+[`Kbucket::can_add()`]: ../dht/struct.Kbucket.html#method.can_add
+[`PackedNode`]: ../dht/struct.PackedNode.html
+[`PING_TIMEOUT`]: ../timeout/constant.PING_TIMEOUT.html
+[`TimeoutQueue`]: ../timeout/struct.TimeoutQueue.html
+*/
+#[derive(Clone)]
+pub struct DhtNode {
+    /// secret key
+    pub sk: SecretKey,
+    /// public key
+    pub pk: PublicKey,
+    /// Close List (contains nodes close to own DHT PK)
+    pub kbucket: Kbucket,
+    /// connected clients cache
+    pub connected_clients: HashMap<PublicKey, Client>,
+
+}
+
+impl DhtNode {
+    /**
+    Create new `DhtNode` instance.
+
+    Note: a new instance generates new DHT public and secret keys.
+
+    DHT `PublicKey` and `SecretKey` are supposed to be ephemeral.
+    */
+    pub fn new() -> Result<DhtNode, Error> {
+        if !crypto_init() {
+            return Err(Error::new(ErrorKind::Other,
+                       "Crypto initialization failed."));
+        }
+
+        let (pk, sk) = gen_keypair();
+        let kbucket = Kbucket::new(KBUCKET_BUCKETS, &pk);
+
+        debug!("Created new DhtNode instance");
+        Ok(DhtNode {
+            sk: sk,
+            pk: pk,
+            kbucket: kbucket,
+            connected_clients: HashMap::new(),
+        })
+    }
+
+    /// add new client entry in connected_clients hashmap
+    pub fn add_client(&mut self, addr: &SocketAddr, packet: &Option<DhtPacket>, tx: Tx) -> Client {
+        let pk = match packet.clone().unwrap() {
+            DhtPacket::PingRequest(ping_request) => ping_request.pk,
+            DhtPacket::PingResponse(ping_response) => ping_response.pk,
+            DhtPacket::NodesRequest(nodes_request) => nodes_request.pk,
+            DhtPacket::NodesResponse(nodes_response) => nodes_response.pk,
+            DhtPacket::DhtRequest(dht_request) => dht_request.spk,
+            _ => unreachable!("Received packet is not a DhtPacket"),
```

Comment:
Also `add_client` must not accept packet to get `pk`. You may create a method to extract client pk from DhtPacket, call it before `add_client` and pass `pk` into `add_client` as computed argument.